### DISCUSSION TITLE
asyncio not compatible with micropython v1.23.0 ?

### DISCUSSION THREAD

**Discussion by kenjican** [micropython/micropython#15442] [ESP32] [answered]
UART with asyncio.StreamWriter, and Streamreader works well before V1.22.2 ,  but not V1.23.0. 

      Code below works before 1.22.2 , sreader could receive 77 bytes within timeout. 
      In V1.23.0 sreader could only receive 8 -10 bytes and not trigger timeout error. Is not it a bug?

```
from machine import UART,Timer,reset
import ujson
import uasyncio as asyncio

swriter = asyncio.StreamWriter(uart, {})
sreader = asyncio.StreamReader(uart)

async def sender():
    uart.read()
    swriter.write(cmdAna)
    await swriter.drain()
    try:
        res = await asyncio.wait_for(sreader.read(77), timeout=1)
        print(res)
        parseAna(res)
    except Exception as e:
        print(e)
        return


async def main(t):
    await asyncio.create_task(sender())


def initCmd(tm):
    t1.init(period=tm * 1000, mode=Timer.PERIODIC, callback=lambda t: asyncio.run(main(1)))
```

**Comment by peterhinch** [comment]
V1.23 is definitely compatible with `asyncio`. Some observations re your code sample:
* The symbol `uart` is not defined. What baudrate is set?
* Issuing the blocking `uart.read()` in an asynchronous function is unwise, especially given the next observation.
* Using a timer to re-run the entire `asyncio` scheduler is a bad idea. It's best to let `asyncio` run continuously and to use an asynchronous function to schedule the sender.
* Whether the `wait_for` will timeout depends entirely on the behaviour of the device at the other end of the serial link.

**Reply by kenjican** [reply]
Thanks for advice. 

-  UART is defined as baud 115200. I forgot to paste these code 
-  uart.read() it to make sure RX buffer is empty. Is there a better solution to clear RX buffer before write?  
-  I change to asyncio.sleep(1) in a while loop. And don't use the timer. like this one: https://github.com/peterhinch/micropython-async/blob/master/v3/as_demos/roundrobin.py  
- I suppose the device is OK, not the cause  

But the same code runs well in V1.22.2, not V1.23.0.  I'll keep testing and figure out why. Thanks for the advice and great asyncio tutorial.

**Reply by peterhinch** [reply, marked as answer]
To ensure the UART input stream is clear, create the `StreamReader` when the application starts. Create a coroutine that reads characters from the stream and discards them. Then call it using `asyncio.wait_for`. The task will time out when the UART stream is empty.

The merit of this approach is that it is non-blocking.

**Comment by kenjican** [comment]
Thanks Peter, testing in V1.24.1 ESP32C3. Replace uart.read() in async def sender():  with sreader.read() .It works. Even I dont know why,problem solved~~
